File: day_8/day8.hs
module Main where

import Data.Foldable (asum)
import Data.Tuple (swap)
import qualified Data.Map as M
import Data.Maybe (mapMaybe)
import Data.List (sort, find, nub)
import           Text.Parsec        hiding (Line, parse)
import           Text.Parsec.String

main :: IO ()
main = do
  input <- parse "input.txt"
  putStrLn "DAY 8"
  putStrLn $ "Solution 1: " ++ show (solve1 input)
  putStrLn $ "Solution 2: " ++ show (solve2 input)

-- * Part 1

solve1 :: Input -> Result
solve1 input = sum $ map (count unique) $ map snd input
  where
    count p = length . filter p
    unique = not . M.null . obvi

-- * Part 2

solve2 :: Input -> Result
solve2 = sum . map ((read . concatMap show) . solveNote)

solveNote :: ([Signal], [Output]) -> [Number]
solveNote (signals, outputs) = decodedOutput
  where
    sortedSignals = map sort signals -- Normalize signals before doing any processing on them
    sortedOutputs = map sort outputs -- Outputs need to match up with normalizes signals

    decodedOutput :: [Number]
    decodedOutput = let mappings = decode sortedSignals
                    in mapMaybe (\output -> M.lookup output mappings) sortedOutputs

decode :: [Signal] -> M.Map Signal Number
decode signals = nonObvious <> obvious
  where
    obvious :: M.Map Signal Number
    obvious = decodeObvious signals

    nonObvious :: M.Map Signal Number
    nonObvious = case decodeNonObvious obvious signals of
                   Nothing -> error "Could not properly decode non obvious numbers .."
                   Just mappings -> mappings

decodeObvious :: [Signal] -> M.Map Signal Number
decodeObvious = foldr (M.union . obvi) mempty

obvi :: Signal -> M.Map Signal Number
obvi s
  | length s == 2 = M.singleton s One
  | length s == 4 = M.singleton s Four
  | length s == 3 = M.singleton s Seven
  | length s == 7 = M.singleton s Eight
  | otherwise = mempty

decodeNonObvious :: M.Map Signal Number -> [Signal] -> Maybe (M.Map Signal Number)
decodeNonObvious mappings signals
  | null unmappedSignals || null unmappedNumbers = pure mappings
  | otherwise =
  -- There are still unmapped letters to map, get going!
  -- Permuate all possible mappings of the remaining letters.
  -- Return a mapping of all the remaining letters such that there are no remaining unmapped letters.

  -- Pick the first mapping that completely solves the current note
  asum $ map (\mapping -> when makesSense (decodeNonObvious mapping signals)) possibleMappings
  where
    -- If the decoding is to work, then the remaining of the unmapped numbers has to have a mapping to all
    -- Signals.
    -- Thus, we start with the numbers instead.
    num = head unmappedNumbers
    possibleMappings = [ M.insert signal num mappings | signal <- unmappedSignals ] -- Among the signals that are left (sus), find a mapping from ONE signal onto `num`

    numbers  = [Zero .. Nine]
    mappedNumbers = M.elems mappings
    unmappedNumbers = filter (`notElem` mappedNumbers) numbers -- I.e. 3,5,6
    unmappedSignals = filter (\signal -> M.notMember signal mappings) signals

    when p ma = ma >>= \a -> if p a then pure a else mempty

-- | Check if a given mapping from signals to numbers make sense
makesSense :: M.Map Signal Number -> Bool
makesSense mappings =
  let correctLenghts = and $ zipWith (==) lenghts (sort $ map (\(num, signal) -> (num, length signal)) $ M.toList reverseMappings)
      validSegments = (length (five \/ two) == 3)                                -- 5,3 should all share 3 equal segments.
                      && (six \\ five == (eight \\ nine))                        -- identifies that a six is indeed a six
                      && (length (one \/ nine) == 2 && length (one \/ six) == 1) -- identifies that a nine is not a six
  in correctLenghts && validSegments
 where
   reverseMappings = invertMap mappings
   (Just one)      = M.lookup One reverseMappings
   (Just two)      = M.lookup Two reverseMappings
   (Just five)     = M.lookup Five reverseMappings
   (Just six)      = M.lookup Six reverseMappings
   (Just eight)    = M.lookup Eight reverseMappings
   (Just nine)     = M.lookup Nine reverseMappings
   lenghts = [(Zero, 6), (One, 2), (Two, 5), (Three, 5), (Four, 4), (Five, 5), (Six, 6), (Seven, 3), (Eight, 7), (Nine, 6)]

   invertMap :: (Ord k, Ord a) => M.Map k a -> M.Map a k
   invertMap = M.fromList . map swap . M.toList

-- | I.e. "abcdefg" `diff` "abcefg" == "d"
(\\) :: String -> String -> [Char]
(\\) s1 s2 = filter (`notElem` s2) s1

-- | I.e. "abc" `sim` "ab" == "ab"
(\/) :: String -> String -> [Char]
(\/) s1 s2 = filter (`elem` s2) s1

-- * Data Types

data Number = Zero | One | Two | Three | Four | Five | Six | Seven | Eight | Nine
  deriving (Eq, Ord, Enum)

instance Show Number where
  show Zero  = "0"
  show One   = "1"
  show Two   = "2"
  show Three = "3"
  show Four  = "4"
  show Five  = "5"
  show Six   = "6"
  show Seven = "7"
  show Eight = "8"
  show Nine  = "9"

type Signal  = [Char]
type Output  = [Char]
type Input   = [([Signal], [Output])]
type Result  = Int

-- * Today's parsing was actually fun and enjoyable. (no sarcasm)

parse :: FilePath -> IO Input
parse input = either (error . show) id <$> parseFromFile parser input

parser :: Parser Input
parser = parseNote `manyTill` eof

parseNote :: Parser ([Signal], [Output])
parseNote = do
  signals <- count 10 (segment <* char ' ')
  char '|'
  char ' '
  output <- count 4 (segment <* (char ' ' <|> char '\n'))
  pure (signals, output)

segment :: Parser [Char]
segment = many letter

File: day_8/day8.py
from itertools import permutations
import pprint
l = []
with open("input.txt") as file:
	for line in file:	
		x = tuple(line.strip().split('|'))
		l.append(x)

def q1(l):
	c = 0
	for t in l:
		x = t[1].split()
		for seg in x:
			l  = len(seg)
			if l == 2 or l == 4 or l == 3 or l == 7:
				c +=1
	return c

def evaluate(x):
	one = [set(item) for item in x if len(item) == 2][0]
	four = [set(item) for item in x if len(item) == 4][0]
	zero = [set(item) for item in x if len(item) == 6 and one.issubset(item) and len(set(item).intersection(four)) == 3][0]
	nine = [set(item) for item in x if len(item) == 6 and one.issubset(item) and len(set(item).intersection(four)) == 4][0]
	six = [set(item) for item in x if len(item) == 6 and not one.issubset(item)][0]
	three = [set(item) for item in x if len(item) == 5 and one.issubset(item) and len(set(item).intersection(six)) == 4][0]
	five = [set(item) for item in x if len(item) == 5 and len(set(item).intersection(six)) == 5][0]
	two = [set(item) for item in x if len(item) == 5 and not one.issubset(item) and len(set(item).intersection(six)) == 4][0]

	return dict(list(map(lambda x: ("".join(list(x[0])),x[1]),[(zero,'0'),(two,'2'),(three,'3'),(five,'5'),(six,'6'),(nine,'9')])))

def q2(l):
	d = {}
	vals = []
	for t in l:
		x = t[0].split()
		output_val = []
		d = evaluate(x)
		output = t[1].split()
		for seg in output:
			if len(seg) == 2:
				output_val.append('1')
			elif len(seg) == 7:
				output_val.append('8')
			elif len(seg) == 4:
				output_val.append('4')
			elif len(seg) == 3:
				output_val.append('7')
			else:
				ps = permutations(seg)
				output_val.append(d[ ["".join(item) for item in ps if "".join(item) in d][0]])
		vals.append(int("".join(output_val)))
	
	return sum(vals)
print("Answer to question 1:",q1(l))
print("Answer to question 2:",q2(l))

